### PR TITLE
[OPIK-471] Improve experiment compare query

### PR DESCRIPTION
**Pull request by thiagohora**
## Details
Attempt to improve experiment compare query. Please see query plan:

Screenshot 2024-12-04 at 16 48 28

Screenshot 2024-12-04 at 16 49 01

## Issues
OPIK-471

### CODE REVIEW

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java [hunk 1]
```diff
@@ -366,93 +427,26 @@ AND id in (
...
-                AND workspace_id = :workspace_id
-                <if(dataset_item_filters)>
-                AND <dataset_item_filters>
-                <endif>
-                ORDER BY id DESC, last_updated_at DESC
-                LIMIT 1 BY id
-            ) AS di
-            INNER JOIN (
-                SELECT
-                    DISTINCT ei.*
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java:383]: Make sure losing this distinct doesn't end up returning the wrong results.

**Review by andrescrz — APPROVED** [review]
The number of consulted parts and granules seems the same and the plan looks pretty similar except for the some extra experiment id filtering, which is what you're aiming for.

Fine to move ahead and check if it provides some performance improvement.

Just make sure the logic of the query is equivalent.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java (hunk 1 above)
Comment (thiagohora) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java:383]: Yeah, we replace the inner join traces with an IN clause, the tests seems to cover it but I can double check

**Comment by thiagohora** [comment]
> The number of consulted parts and granules seems the same and the plan looks pretty similar except for the some extra experiment id filtering, which is what you're aiming for.
> 
> Fine to move ahead and check if it provides some performance improvement.
> 
> Just make sure the logic of the query is equivalent.

The improvements are for experiments and trace filters; I hope this helps.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java (hunk 1 above)
Comment (thiagohora) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java:383]: Confirmed the tests verify this, by checking number of rows and and their values

**Pull request merged**